### PR TITLE
[TH2-2675] Implemented new serialization mechanism

### PR DESCRIPTION
**Pull request by std000**

### CODE REVIEW

File: cradle-core/src/main/java/com/exactpro/cradle/serialization/EventMessageIdDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	{
+		if (bytes == null || bytes.length == 0)
+			return null;
+
+		try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
+			 DataInputStream dis = new DataInputStream(bais))
+		{
+			byte version = dis.readByte();
+			if (version != VERSION)
+				throw new IOException("Unsupported data format version - "+version);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/serialization/EventMessageIdDeserializer.java:47]: Expected version would be useful here as well.

File: cradle-core/src/main/java/com/exactpro/cradle/serialization/Serialization.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.cradle.serialization;
+
+public class Serialization {
+
+	public static final String INVALID_MAGIC_NUMBER_FORMAT = "Invalid magic number for %s. Got: %s. Expected %s. Probably received inconsistent data.";
+	public static final String NOT_SUPPORTED_PROTOCOL_FORMAT = "Not supported %s protocol version. Supported %s. Got: %s.";
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/serialization/Serialization.java:22]: This should be used for errors where VERSION is invalid, I think.

File: cradle-core/src/main/java/com/exactpro/cradle/serialization/SerializationUtils.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+			buffer.put(body);	
+		}
+	}
+
+	static void printShortString(String value, ByteBuffer buffer, String paramName) throws SerializationException {
+		if (value == null) {
+			value = "";
+		}
+		if (value.length() > 65535) {
+			throw new SerializationException(paramName + " is too big. Expected length [0-65535]");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/serialization/SerializationUtils.java:42]: 65535 should be a constant.

File: cradle-core/src/test/java/com/exactpro/cradle/utils/SerializationMessageTest.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+		builder.setContent("Message".repeat(10).getBytes(StandardCharsets.UTF_8));
+		StoredMessage build = builder.build();
+		MessageSerializer serializer = new MessageSerializer();
+		try {
+			serializer.serialize(build);
+		} catch (SerializationException e) {
+			Assert.assertTrue(true);
+			return;
+		}
+		Assert.fail("Should be an error when stream name is more than 65536");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/test/java/com/exactpro/cradle/utils/SerializationMessageTest.java:104]: `...when stream name is longer than...`

File: cradle-core/src/main/java/com/exactpro/cradle/serialization/MessageDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		Direction[] values = Direction.values();
+		if (values.length > ordinal && ordinal >= 0)
+			return values[ordinal];
+		throw new SerializationException(String.format("Invalid ordinal for enum (Direction): %d. Values: [0-%d]",
+				ordinal, values.length - 1));
+	}
+
+	private StoredMessageId readMessageId(ByteBuffer buffer) throws SerializationException {
+		String stream = readShortString(buffer);
+		Direction direction = getDirection(buffer.get());
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/serialization/MessageDeserializer.java:122]: I'm sure you can omit saving of this for all messages in batch. These two values should be saved only once per batch.

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: cradle-core/src/main/java/com/exactpro/cradle/serialization/MessageSerializer.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+import static com.exactpro.cradle.serialization.Serialization.MessageBatchConst.*;
+import static com.exactpro.cradle.serialization.SerializationUtils.printBody;
+import static com.exactpro.cradle.serialization.SerializationUtils.printInstant;
+import static com.exactpro.cradle.serialization.SerializationUtils.printShortString;
+import static com.exactpro.cradle.serialization.SerializationUtils.printString;
+
+public class MessageSerializer {
+
+	public byte[] serializeBatch(Collection<StoredMessage> batch) throws SerializationException {
+		SerializationBatchSizes messageBatchSizes = calculateMessageBatchSize(batch);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/serialization/MessageSerializer.java:37]: Please use `calculateMessageBatchSize()` method in calculating current batch size in `StoredMessageBatch` where it checks if message being added won't overload the batch.

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: cradle-core/src/test/java/com/exactpro/cradle/utils/SerializationMessageTest.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+		builder.setContent("Message".repeat(10).getBytes(StandardCharsets.UTF_8));
+		StoredMessage build = builder.build();
+		MessageSerializer serializer = new MessageSerializer();
+		try {
+			serializer.serializeBatch(Collections.singleton(build));
+		} catch (SerializationException e) {
+			Assert.assertTrue(true);
+			return;
+		}
+		Assert.fail("Should be an error when stream name is longer than 65536");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/test/java/com/exactpro/cradle/utils/SerializationMessageTest.java:107]: Use `SerializationUtils.USHORT_MAX_VALUE here?

**Comment by Vladimir-Panarin-Exactpro** [comment]
It worth adding a test when multiple messages with small content and long stream names are added to batch. Previously the batch won't find this problematic. Now it should prohibit adding messages due to actual message size calculated.
Limit for batch can be specified really low for this test.

File: cradle-core/src/main/java/com/exactpro/cradle/serialization/MessageSerializer.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+		if (messageBatchSizes == null) {
+			messageBatchSizes = MessagesSizeCalculator.calculateMessageBatchSize(batch);
+		}
+		
+		buffer.putInt(MESSAGE_BATCH_MAGIC);
+		buffer.put(MESSAGE_PROTOCOL_VER);
+
+		StoredMessageId msgId = batch.isEmpty() ? new StoredMessageId("", Direction.FIRST, 0L) 
+				: batch.iterator().next().getId();
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/serialization/MessageSerializer.java:54]: batch.getId() looks sufficient.

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: cradle-core/src/main/java/com/exactpro/cradle/serialization/MessageSerializer.java (hunk 8 above)
Comment (std000) [reply cradle-core/src/main/java/com/exactpro/cradle/serialization/MessageSerializer.java:54]: batch - is StoredMessage collection, not StoredMessageBatch

**Comment by MrJonniK** [comment]
Merged with ver_3_1_dev in commit https://github.com/th2-net/cradleapi/commit/8bc98a7dbe35d7e0b229e2e66b492c878bcead34

**Pull request closed without merging**